### PR TITLE
Restructure storage db

### PR DESCRIPTION
**Pull request by ecdsa**
- perform json serialization/deserialization in json_db
 - use attr.s instead of namedtuples for channel config
 - use StorageDict that keeps track of modifications
 - storage upgrade:
    flatten 'addresses'
    convert lists to dict
    convert FeeUpdate to dict

### CODE REVIEW

File: electrum/lnutil.py [hunk 1]
```diff
@@ -37,74 +38,74 @@
...
-                'ctns': {int(LOCAL): self.ctns[LOCAL],
-                         int(REMOTE): self.ctns[REMOTE]}}
-
-
-ChannelConstraints = namedtuple("ChannelConstraints", ["capacity", "is_initiator", "funding_txn_minimum_depth"])
+@attr.s
+class ChannelConstraints(StoredAttr):
+    capacity = attr.ib(int)
+    is_initiator = attr.ib(bool)
+    funding_txn_minimum_depth = attr.ib(int)
```
Comment (SomberNight) [inline electrum/lnutil.py]: missing `type=`

File: electrum/lnhtlc.py [hunk 2]
```diff
@@ -212,12 +191,15 @@ def store_local_update_raw_msg(self, raw_update_msg: bytes, *, is_commitment_sig
             ctn_idx = self.ctn_latest(REMOTE)
         else:
             ctn_idx = self.ctn_latest(REMOTE) + 1
-        if ctn_idx not in self.log['unacked_local_updates2']:
-            self.log['unacked_local_updates2'][ctn_idx] = []
-        self.log['unacked_local_updates2'][ctn_idx].append(raw_update_msg)
+        ctn_idx = str(ctn_idx)
+        l = self.log['unacked_local_updates2'].get(ctn_idx, [])
```
[the remark this answers is not in the document]
Comment (ecdsa) [reply electrum/lnhtlc.py]: Actually that's what they do. 
The str conversion is unnecessary, it was written for an earlier version.

File: electrum/json_db.py [hunk 3]
```diff
@@ -29,30 +29,138 @@
...
+            v = dict((k, UpdateAddHtlc(*x)) for k, x in value.items())
+        elif key == 'fee_updates':
+            v = dict((k, FeeUpdate(**x)) for k, x in value.items())
+        elif key == 'tx_fees':
+            v = dict((k, TxFeesValue(*x)) for k, x in value.items())
+        elif key == 'prevout_by_scripthash':
+            v = dict((k, {(prevout, value) for (prevout, value) in x}) for k, x in value.items())
+        elif key == 'buckets':
+            v = dict((k, ShachainElement(bfh(x[0]), int(x[1]))) for k, x in value.items())
+        else:
```
[the remark this answers is not in the document]
Comment (ecdsa) [reply electrum/json_db.py]: We can use the complete path, it is passed to StorageDict In the jsonpatch branch.
I did not want to add that to the current commit, but maybe we should.

File: electrum/json_db.py [hunk 4]
```diff
@@ -29,30 +29,138 @@
...
+
+
 class TxFeesValue(NamedTuple):
     fee: Optional[int] = None
     is_calculated_by_us: bool = False
     num_inputs: Optional[int] = None
 
 
+class StorageDict(dict):
+
```
[the remark this answers is not in the document]
Comment (ecdsa) [reply electrum/json_db.py]: indeed, `pop` and `get` should be overridden

File: electrum/json_db.py [hunk 5]
```diff
@@ -741,18 +909,11 @@ def list_verified_tx(self):
...
-                           conf=None,
-                           timestamp=timestamp,
-                           txpos=txpos,
-                           header_hash=header_hash)
+        return self.verified_tx.get(txid)
 
     @modifier
     def add_verified_tx(self, txid, info):
-        self.verified_tx[txid] = (info.height, info.timestamp, info.txpos, info.header_hash)
+        self.verified_tx[txid] = info
```
[the remark this answers is not in the document]
Comment (ecdsa) [reply electrum/json_db.py]: ok, I will leave those unchanged. 
Longer term I think `conf` should not be part of `TxMinedInfo`

File: electrum/lnhtlc.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ecdsa) [reply electrum/lnhtlc.py]: that was already the case for `__getitem__` but not for `get`.
I overloaded it

File: electrum/json_db.py [hunk 6]
```diff
@@ -29,30 +29,138 @@
...
 NEW_SEED_VERSION = 11       # electrum versions >= 2.0
-FINAL_SEED_VERSION = 22     # electrum >= 2.7 will set this to prevent
+FINAL_SEED_VERSION = 23     # electrum >= 2.7 will set this to prevent
                             # old versions from overwriting new format
 
 
 JsonDBJsonEncoder = util.MyEncoder
 
 
+def decodeAll(d, local):
```
[the remark this answers is not in the document]
Comment (ecdsa) [reply electrum/json_db.py]: got rid of it now, everything is performed in the class

File: electrum/json_db.py [hunk 7]
```diff
@@ -53,6 +57,122 @@ class TxFeesValue(NamedTuple):
...
+        self.path = path
+        # recursively convert dicts to storagedict
+        for k, v in list(data.items()):
+            self.__setitem__(k, v)
+
+    def convert_key(self, key):
+        # convert int, HTLCOwner to str
+        return str(int(key)) if isinstance(key, int) else key
+
+    def __setitem__(self, key, v):
```
Comment (SomberNight) [inline electrum/json_db.py]: storage.write() takes db.lock;
should StorageDict.__setitem__ and similar methods not take db.lock?

File: electrum/lnutil.py [hunk 8]
```diff
@@ -113,7 +114,10 @@ class ScriptHtlc(NamedTuple):
 
 
 # FIXME duplicate of TxOutpoint in transaction.py??
-class Outpoint(NamedTuple("Outpoint", [('txid', str), ('output_index', int)])):
+@attr.s
+class Outpoint(StoredAttr):
+    txid = attr.ib(str)
+    output_index = attr.ib(int)
```
Comment (SomberNight) [inline electrum/lnutil.py]: `type=` ? :P

**Pull request closed without merging**